packages/data/src/components/with-select/index.js

```diff
@@ -63,35 +63,48 @@ const withSelect = ( mapSelectToProps ) => createHigherOrderComponent( ( Wrapped
 		}
 
 		shouldComponentUpdate( nextProps, nextState ) {
-			const hasPropsChanged = ! isShallowEqual( this.props.ownProps, nextProps.ownProps );
+			// Cycle subscription if registry changes.
+			const hasRegistryChanged = nextProps.registry !== this.props.registry;
+			if ( hasRegistryChanged ) {
+				this.unsubscribe();
+				this.subscribe( nextProps.registry );
```

Comment:
I don't like that we're doing this magic in `shouldComponentUpdate`. It doesn't seem like the best function for this kind of side effects.  Ideally it's `componentWillReceiveProps` but this one has been deprecated, so not sure what's the "good" approach for these at the moment.

I see `shouldComponentUpdate` as a function that can be called multiple times by React even without rendering (especially in async) as it's supposed to be a pure function just returning a boolean.

Reply:
My general approach with these components has been to happily throw out all best practices if it means performance could suffer even marginally in the individual case by following them.

If there's an unexpected behavior that could occur in how we're implementing (even "number of calls"), we should have a unit test to cover it in whatever fix need to be made.

Reply:
Makes sense 👍